File: Labs/week6/A_merge_sort_inversions.py
cnt = 0

def merge_sort(arr, l, r):
    global cnt
    if l >= r:
        return
    mid = (l + r) // 2
    merge_sort(arr, l, mid)
    merge_sort(arr, mid + 1, r)
    temp = []
    i, j = l, mid + 1
    while i <= mid and j <= r:
        if arr[i] <= arr[j]:
            temp.append(arr[i])
            i += 1
        else:
            temp.append(arr[j])
            cnt += mid - i + 1
            j += 1
    while i <= mid:
        temp.append(arr[i])
        i += 1
    while j <= r:
        temp.append(arr[j])
        j += 1
    for k in range(len(temp)):
        arr[l + k] = temp[k]

n = int(input())
arr = list(map(int, input().split()))
merge_sort(arr, 0, n - 1)
print(f"count = {cnt}")
print(' '.join(map(str, arr)))

File: Labs/week6/A_merge_sort_inversions.cpp
#include <iostream>
using namespace std;

int q[105], temp[105], cnt = 0;

void merge_sort(int l, int r) {
    if (l >= r) return;
    int mid = (l + r) / 2, i = l, j = mid + 1, k = 0;
    merge_sort(l, mid);
    merge_sort(mid + 1, r);
    while (i <= mid && j <= r) {
        if (q[i] <= q[j]) temp[k++] = q[i++];
        else temp[k++] = q[j++], cnt += mid - i + 1;
    }
    while (i <= mid) temp[k++] = q[i++];
    while (j <= r) temp[k++] = q[j++];
    for (i = l, j = 0; i <= r; i++, j++) q[i] = temp[j];
}

int main() {
    int n;
    cin >> n;
    for (int i = 0; i < n; i++) cin >> q[i];
    merge_sort(0, n - 1);
    cout << "count = " << cnt << endl;
    for (int i = 0; i < n; i++) cout << q[i] << " ";
    cout << endl;
    return 0;
}
